prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+
+    for i in range(1, mar + 1):
+        z, y, x = np.where(lav == i)
+        curr_z = z.max()
+        if curr_z > z_max:
+            curr_z = z_max
+            coords = (z, y, x)
+
+        if curr_z == z_max:
+            if volumes[i] > volumes[max_marker]:
+                # curr_z = z_max
+                coords = (z, y, x)
+
+        err_lung[coords] = 0
+    return err_lung
+
+
+def costal_surface(lung, z_coords, max_coor, combined):
+    """ The costal surface.
+
+    The convexity is determined by comparing the costal lung surface
+    in axial slices to the convex hull of this costal lung surface.
+
+    Args:
+        lung: cumulative x-position in lung
+        z_coords: z-coordinates in lung
+        max_coor: threshold for the сostal surface for the lung
+        combined (np.ndarray[bool]): segmented lung (left or right)
+
+    """
+    erroneus = np.zeros(lung.shape)
+    for coord in np.unique(z_coords):
+        if max_coor == .8:
+            ROI = (lung[coord] < max_coor) * (lung[coord] != 0)
+        else:
+            ROI = (lung[coord] > max_coor) * (lung[coord] != 0)
+
+        erroneus[coord] = skimage.morphology.convex_hull_object(ROI) - ROI
+
+    erroneus = erroneus * (1 - (combined != 0))
+    erroneus_erosion = scipy.ndimage.morphology.binary_erosion(erroneus, structure=skimage.morphology.ball(4.5))
+    erroneus_erosion = scipy.ndimage.morphology.binary_dilation(erroneus_erosion,
+                                                                structure=skimage.morphology.ball(4.5))
+
+    return lung, erroneus, erroneus_erosion
+
+
+def detection_lung_error(patient):
+    """ Automatic error detection.
+
+    The function aggregates cumulation, costal_surface, ventricular_extraction.
+    Args:
+        patient
+
+    Returns:
+        lung_l - the mask of left lung
+        lung_r - the mask of right lung
+        trachea - the mask of trachea
+
+    """
+
+    combined, trachea = conventional_lung_segmentation(patient)
+
+    lung_right = ((combined.copy() == 1)
+                  | (combined.copy() == 3)
+                  | (combined.copy() == 5)
+                  | (combined.copy() == 7)
+                  | (combined.copy() == 9)
+                  | (combined.copy() == 11)
+                  | (combined.copy() == 13)
+                  | (combined.copy() == 15))
+
+    lung_left = ((combined.copy() == 2)
+                 | (combined.copy() == 4)
+                 | (combined.copy() == 6)
+                 | (combined.copy() == 8)
+                 | (combined.copy() == 10)
+                 | (combined.copy() == 12)
+                 | (combined.copy() == 14))
+
+    lung_right_c = lung_right.astype(float)
+    lung_left_c = lung_left.astype(float)
+
+    lung_right_comm, z_coords_r = cumulation(lung_right_c)
+    lung_left_comm, z_coords_l = cumulation(lung_left_c)
+
+    lung_left_1, er_l, erroneus_erosion_left = costal_surface(lung_left_comm, z_coords_l, .2, lung_left_c)
```

Comment:
Can you explain the magic numbers .2 and .8 using meaningful variable names, please? :)

Reply:
It was in https://www.ncbi.nlm.nih.gov/pubmed/19673192.
This values are needed to leave out the mediastinal region of the lungs for which the convexity constraint does not hold:
x > 0.2 for the left lung;
x < 0.8 for the right lung.
I made a mistake in variable names, and there is a small confusion :( However, it has already been fixed in a new version, which hasn't been send yet. This is all in the progress!